core/server/common/src/main/java/alluxio/util/ParallelZipUtils.java

```
@@ -0,0 +1,181 @@
+    ExecutorService executor = ExecutorServiceFactories.fixedThreadPool(
+        "parallel-zip-compress-pool", poolSize).create();
+    ParallelScatterZipCreator parallelScatterZipCreator = new ParallelScatterZipCreator(executor);
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(backupPath);
+         BufferedOutputStream bufferedOutputStream =
+             new BufferedOutputStream(fileOutputStream);
+         ZipArchiveOutputStream zipArchiveOutputStream =
+             new ZipArchiveOutputStream(bufferedOutputStream)) {
+      try (final Stream<Path> stream = Files.walk(dirPath)) {
+        for (Path subPath : stream.collect(toList())) {
+          if (Thread.interrupted()) {
+            throw new InterruptedException();
+          }
+
+          File file = subPath.toFile();
+
+          final InputStreamSupplier inputStreamSupplier = () -> {
+            try {
+              return new FileInputStream(file);
+            } catch (FileNotFoundException e) {
+              LOG.warn("Can't find file when parallel zip, path = {}", subPath);
+              return new NullInputStream(0);
+            }
+          };
+
+          String entryName = dirPath.relativize(subPath).toString();
+          if (file.isDirectory()) {
+            entryName += File.separator;
+          }
+
+          ZipArchiveEntry zipArchiveEntry = new ZipArchiveEntry(entryName);
+          zipArchiveEntry.setMethod(ZipArchiveEntry.DEFLATED);
+
+          parallelScatterZipCreator.addArchiveEntry(zipArchiveEntry, inputStreamSupplier);
+        }
+      }
+
+      parallelScatterZipCreator.writeTo(zipArchiveOutputStream);
+      zipArchiveOutputStream.flush();
+      fileOutputStream.getFD().sync();
+    } catch (ExecutionException e) {
+      LOG.error("Parallel compress rocksdb fail", e);
+      ExecutorServiceUtils.shutdownAndAwaitTermination(executor);
+      FileUtils.deletePathRecursively(backupPath);
+      throw new RuntimeException(e);
```

Actually I wonder if failing to backup should crash the master with a RuntimeException, because we might not have to. @tcrain WDYT? I feel the master can keep going or just fall back to a single-threaded backup, or just ignore the error and retry next time.

---

As an utility class, you can just wrap whatever error here into IOException and let the caller decide. The caller will determine if a crash or retry is applicable.